### PR TITLE
JBIG2Decoder implementation

### PR DESCRIPTION
**Pull request by kucjac**
This PR contains JBIG2Decoder implementation. All the package structures are defined in the internal package jbig2.

The package requires the row stride of the image to be taken into account.

Currently the Decode [1.0 0.0] parameter is treated by the JBIG2Encoder - so that it flips the meaning of the bits '0' and '1' b/w.

---
This change is [Reviewable](https://reviewable.io/reviews/unidoc/unipdf/67)

### CODE REVIEW

File: internal/jbig2/segments/generic-refinement-region.go [hunk 1]
```diff
@@ -0,0 +1,1020 @@
...
+		c4 = (((c4 << 1) | 0x01&int16(w4>>7)) & 0x07)
+		c5 = int16(bit)
+
+		if (x-g.ReferenceDX)%8 == 5 {
+			if ((x-g.ReferenceDX)/8)+1 >= g.ReferenceBitmap.RowStride {
+				w1 = 0
+				w2 = 0
+				w3 = 0
+			} else {
+				if currentLine >= 1 && (currentLine-1) < g.ReferenceBitmap.Height {
```
Comment (adrg) [inline internal/jbig2/segments/generic-refinement-region.go]: Could set `w1`, `w2` and `w3` before each if statement and remove the else branches.

File: internal/jbig2/segments/generic-refinement-region.go [hunk 2]
```diff
@@ -0,0 +1,1020 @@
...
+	if b.GetPixel(x, y) {
+		return 1
+	}
+	return 0
+}
+
+func (g *GenericRefinementRegion) overrideAtTemplate0(context, x, y, result, minorX int) int {
+	if g.grAtOverride[0] {
+		context &= 0xfff7
+
```
Comment (adrg) [inline internal/jbig2/segments/generic-refinement-region.go]: Remove empty line for consistency with the code block below.

File: internal/jbig2/segments/generic-refinement-region.go [hunk 3]
```diff
@@ -0,0 +1,1020 @@
...
+	defer func() {
+		if err == nil {
+			common.Log.Trace("[GENERIC-REF-REGION] parsing header finishid in: %d ns", time.Since(ts).Nanoseconds())
+		} else {
+			common.Log.Trace("[GENERIC-REF-REGION] parsing header failed: %s", err)
+		}
+	}()
+
+	if err = g.RegionInfo.parseHeader(); err != nil {
+		return
```
Comment (adrg) [inline internal/jbig2/segments/generic-refinement-region.go]: Please add explicit returns throughout the method.

File: internal/jbig2/segments/generic-refinement-region.go [hunk 4]
```diff
@@ -0,0 +1,1020 @@
...
+
+// setParameters sets the parameters for the Generic Refinemenet Region.
+func (g *GenericRefinementRegion) setParameters(
+	cx *arithmetic.DecoderStats, arithmDecoder *arithmetic.Decoder,
+	grTemplate int8, regionWidth, regionHeight int,
+	grReference *bitmap.Bitmap, grReferenceDX, grReferenceDY int,
+	isTPGRon bool, grAtX []int8, grAtY []int8,
+) {
+	common.Log.Trace("[GENERIC-REF-REGION] setParameters")
+	defer func() {
```
Comment (adrg) [inline internal/jbig2/segments/generic-refinement-region.go]: Could just add the log at the bottom of the method as there are no other exit points. No defer function seems to be needed here.

File: internal/jbig2/segments/generic-region.go [hunk 5]
```diff
@@ -0,0 +1,1117 @@
...
+	arithDecoder *arithmetic.Decoder
+	cx           *arithmetic.DecoderStats
+	mmrDecoder   *mmr.Decoder
+}
+
+// NewGenericRegion creates new GenericRegion segment.
+func NewGenericRegion(
+	r reader.StreamReader,
+) *GenericRegion {
+	g := &GenericRegion{
```
Comment (adrg) [inline internal/jbig2/segments/generic-region.go]: The `g` variable is not needed.

File: internal/jbig2/segments/generic-region.go [hunk 6]
```diff
@@ -0,0 +1,1117 @@
...
+	if g.IsMMREncoded {
+		// MMR Decoder Call
+		if g.mmrDecoder == nil {
+			g.mmrDecoder, err = mmr.New(
+				g.r,
+				g.RegionSegment.BitmapWidth, g.RegionSegment.BitmapHeight,
+				g.DataOffset, g.DataLength,
+			)
+			if err != nil {
+				return
```
Comment (adrg) [inline internal/jbig2/segments/generic-region.go]: Please add explicit returns for the function.

File: internal/jbig2/segments/generic-region.go [hunk 7]
```diff
@@ -0,0 +1,1117 @@
...
+	b, err = g.r.ReadBit()
+	if err != nil {
+		return err
+	}
+	if b == 1 {
+		g.IsMMREncoded = true
+	}
+
+	if !g.IsMMREncoded {
+		var numberOfGbAt int
```
Comment (adrg) [inline internal/jbig2/segments/generic-region.go]: Could initialize `numberOfGbAt` with `1` and remove the else branch. Also inside the if statement, the else branch can be eliminated by assigned `4` to `numberOfGbAt`.

File: internal/jbig2/segments/halftone-segment.go [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+			); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func newHalftoneRegion(r *reader.Reader) *HalftoneRegion {
+	hr := &HalftoneRegion{
```
Comment (adrg) [inline internal/jbig2/segments/halftone-segment.go]: No need for the `hr` variable.

File: internal/jbig2/segments/region.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+	// Region segment bitmap Y location, 7.4.1.4
+	YLocation int
+
+	// Region segment flags, 7.4.1.5
+	CombinaionOperator bitmap.CombinationOperator
+}
+
+// NewRegionSegment creates new Region segment model.
+func NewRegionSegment(r reader.StreamReader) *RegionSegment {
+	rs := &RegionSegment{r: r}
```
Comment (adrg) [inline internal/jbig2/segments/region.go]: No need for the `rs` variable.

File: internal/jbig2/segments/symbol-dictionary.go [hunk 10]
```diff
@@ -0,0 +1,1127 @@
...
+			}
+		}
+
+		if exRunLength != 0 {
+			for index := exportIndex; index < exportIndex+exRunLength; index++ {
+				exportFlags[index] = currentExportFlag
+			}
+		}
+
+		if currentExportFlag == 0 {
```
Comment (adrg) [inline internal/jbig2/segments/symbol-dictionary.go]: Could assign `0` to `currentExportFlag` before the if statement and remove the else branch.

File: internal/jbig2/segments/symbol-dictionary.go [hunk 11]
```diff
@@ -0,0 +1,1127 @@
...
+				return fmt.Errorf("related segment with index: %d getDictionary failed. %s", h.SegmentNumber, err.Error())
+			}
+			s.importSymbols = append(s.importSymbols, relatedDict...)
+			s.numberOfImportedSymbols += dict.numberOfExportedSymbols
+		}
+	}
+	return nil
+}
+
+func (s *SymbolDictionary) setAtPixels() error {
```
Comment (adrg) [inline internal/jbig2/segments/symbol-dictionary.go]: Could use a variable here so not to repeat code. Something like:

```
func (s *SymbolDictionary) setAtPixels() error {
	if !s.isHuffmanEncoded {
		index := 1
		if s.sdTemplate == 0 {
			index = 4
		}

		if err := s.readAtPixels(index); err != nil {
			return err
		}
	}
	return nil
}
```

**Comment by adrg** [comment]
___
*[internal/jbig2/document.go, line 128 at r10](https://reviewable.io/reviews/unidoc/unipdf/67#-LgeGJsP1PbCVlqQddW9-r10-128:-Lio4iwAAxEiHB0Gd4uh:bxs4wrv) ([raw file](https://github.com/unidoc/unipdf/blob/a40a01f66eda971a4d912e53538d07e95802fa55/internal/jbig2/document.go#L128)):*
<details><summary><i>Previously, kucjac (Jacek Kucharczyk) wrote…</i></summary><blockquote>

This function return no values, and current use of the 'return' statement, is to end the function fast if the condition is not fulfilled.
</blockquote></details>

You are right, sorry. I must have been looking at another method.

**Comment by gunnsth** [comment]
@kucjac I've integrated this with our build system, such that it runs all the integration tests on build (Jenkinsfile update). Changed the environment variable to UNIDOC_JBIG2_TESTDATA for consistency with other integration tests requiring external test data.
Looking at test output, a lot of the extracted images, but a significant fraction appears to have some row stride issues, i.e. content going across lines, like:
image
(003685_31_0.pdf).

Also changed the base branch to development as all new things going in there  prior to release.

**Pull request merged**